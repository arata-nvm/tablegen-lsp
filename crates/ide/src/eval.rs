use std::sync::Arc;

use ecow::EcoString;

use syntax::ast;
use syntax::ast::AstNode;
use syntax::ast::SourceFile;

use crate::db::SourceDatabase;
use crate::file::FileId;
use crate::symbol_map::{Class, SymbolMap};

#[salsa::query_group(EvalDatabaseStorage)]
pub trait EvalDatabase: SourceDatabase {
    fn symbol_map(&self, file_id: FileId) -> Arc<SymbolMap>;
}

fn symbol_map(db: &dyn EvalDatabase, file_id: FileId) -> Arc<SymbolMap> {
    let parse = db.parse(file_id);
    let source_file = ast::SourceFile::cast(parse.syntax_node()).unwrap();
    Arc::new(eval(source_file))
}

fn eval(source_file: SourceFile) -> SymbolMap {
    let mut ctx = EvalCtx::new();
    source_file.eval(&mut ctx);
    ctx.symbol_map
}

#[derive(Debug, Default)]
pub struct EvalCtx {
    symbol_map: SymbolMap,
}

impl EvalCtx {
    pub fn new() -> Self {
        Self::default()
    }
}

pub trait Eval {
    type Output;
    fn eval(self, ctx: &mut EvalCtx) -> Option<Self::Output>;
}

impl Eval for ast::SourceFile {
    type Output = ();
    fn eval(self, ctx: &mut EvalCtx) -> Option<Self::Output> {
        self.statement_list()?.eval(ctx);
        Some(())
    }
}

impl Eval for ast::StatementList {
    type Output = ();
    fn eval(self, ctx: &mut EvalCtx) -> Option<Self::Output> {
        for stmt in self.statements() {
            stmt.eval(ctx);
        }
        Some(())
    }
}

impl Eval for ast::Statement {
    type Output = ();
    fn eval(self, ctx: &mut EvalCtx) -> Option<Self::Output> {
        match self {
            ast::Statement::Class(class) => class.eval(ctx),
            _ => unimplemented!(),
        }
    }
}

impl Eval for ast::Class {
    type Output = ();
    fn eval(self, ctx: &mut EvalCtx) -> Option<Self::Output> {
        let name = self.name()?.eval(ctx)?;
        let class = Class::new(name);
        ctx.symbol_map.add_class(class);
        Some(())
    }
}

impl Eval for ast::Identifier {
    type Output = EcoString;
    fn eval(self, _ctx: &mut EvalCtx) -> Option<Self::Output> {
        self.value()
    }
}
