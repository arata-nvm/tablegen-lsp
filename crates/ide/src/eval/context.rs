use ecow::{eco_format, EcoString};
use syntax::parser::TextRange;

use crate::{
    file_system::{FileId, FileRange},
    handlers::diagnostics::Diagnostic,
    symbol_map::{symbol::SymbolId, variable::VariableId, SymbolMap},
};

use super::{scope::Scopes, EvalDatabase, Evaluation};

pub struct EvalCtx<'a> {
    pub db: &'a dyn EvalDatabase,
    pub file_trace: Vec<FileId>,
    pub symbol_map: SymbolMap,
    pub diagnostics: Vec<Diagnostic>,
    pub scopes: Scopes,
    pub anonymous_def_index: u32,
}

impl<'a> EvalCtx<'a> {
    pub fn new(db: &'a dyn EvalDatabase, root_file: FileId) -> Self {
        Self {
            db,
            file_trace: vec![root_file],
            symbol_map: SymbolMap::default(),
            diagnostics: Vec::new(),
            scopes: Scopes::default(),
            anonymous_def_index: 0,
        }
    }

    pub fn current_file_id(&self) -> FileId {
        *self.file_trace.last().expect("file_trace is empty")
    }

    pub fn push_file(&mut self, file_id: FileId) {
        self.file_trace.push(file_id);
    }

    pub fn pop_file(&mut self) {
        self.file_trace.pop().expect("file_trace is empty");
    }

    pub fn resolve_id(&self, name: &EcoString) -> Option<SymbolId> {
        if let Some(def_id) = self.symbol_map.find_def(name) {
            return Some(def_id.into());
        }
        if let Some(symbol_id) = self.scopes.find_local(name) {
            return Some(symbol_id);
        }
        None
    }

    pub fn resolve_id_in_current_scope(&self, name: &EcoString) -> Option<VariableId> {
        self.scopes.find_variable_in_current_scope(name)
    }

    pub fn error(&mut self, range: TextRange, message: impl Into<String>) {
        let file = self.current_file_id();
        self.diagnostics
            .push(Diagnostic::new(FileRange::new(file, range), message));
    }

    pub fn finish(self) -> Evaluation {
        Evaluation {
            symbol_map: self.symbol_map,
            diagnostics: self.diagnostics,
        }
    }

    pub fn next_anonymous_def_name(&mut self) -> EcoString {
        let index = self.anonymous_def_index;
        self.anonymous_def_index += 1;
        eco_format!("anonymous_{index}")
    }
}
