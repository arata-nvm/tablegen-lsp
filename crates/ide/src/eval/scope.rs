use std::collections::HashMap;

use ecow::EcoString;

use crate::symbol_map::{
    class::ClassId,
    def::DefId,
    record::Record,
    symbol::SymbolId,
    variable::{Variable, VariableId},
    SymbolMap,
};

#[derive(Debug)]
pub struct Scopes {
    scopes: Vec<Scope>,
}

impl Default for Scopes {
    fn default() -> Self {
        Self {
            scopes: vec![Scope::new(ScopeKind::Root)],
        }
    }
}

impl Scopes {
    pub fn push(&mut self, kind: ScopeKind) {
        self.scopes.push(Scope::new(kind));
    }

    pub fn pop(&mut self) -> Scope {
        self.scopes.pop().expect("scope is empty")
    }

    pub fn current_record_id(&self) -> SymbolId {
        self.scopes
            .iter()
            .rev()
            .find_map(|scope| scope.record_id())
            .expect("scope is empty")
    }

    pub fn current_record(&self) -> &Record {
        self.scopes
            .iter()
            .rev()
            .find_map(|scope| scope.record())
            .expect("scope is empty")
    }

    pub fn current_record_mut(&mut self) -> &mut Record {
        self.scopes
            .iter_mut()
            .rev()
            .find_map(|record| record.record_mut())
            .expect("scope is empty")
    }

    pub fn add_variable(&mut self, symbol_map: &mut SymbolMap, variable: Variable) {
        let name = variable.name.clone();
        let id = symbol_map.add_variable(variable);
        let current_scope = self.scopes.last_mut().expect("scope is empty");
        current_scope.name_to_variable.insert(name, id);
    }

    pub fn find_local(&self, name: &EcoString) -> Option<SymbolId> {
        for scope in self.scopes.iter().rev() {
            if let Some(id) = scope.find_variable(name) {
                return Some(id.into());
            }
            if let Some(record) = scope.record() {
                if let Some(field_id) = record.find_field(name) {
                    return Some(field_id.into());
                }
                if let Some(template_arg_id) = record.find_template_arg(name) {
                    return Some(template_arg_id.into());
                }
            }
        }
        None
    }

    pub fn find_variable_in_current_scope(&self, name: &EcoString) -> Option<VariableId> {
        let scope = self.scopes.last().expect("scope is empty");
        scope.find_variable(name)
    }
}

#[derive(Debug)]
pub struct Scope {
    pub kind: ScopeKind,
    pub name_to_variable: HashMap<EcoString, VariableId>,
}

#[derive(Debug)]
pub enum ScopeKind {
    Root,
    Class(ClassId, Record),
    Def(DefId, Record),
    Foreach(EcoString, VariableId),
}

impl Scope {
    pub fn new(kind: ScopeKind) -> Self {
        Self {
            kind,
            name_to_variable: HashMap::new(),
        }
    }

    pub fn record_id(&self) -> Option<SymbolId> {
        match self.kind {
            ScopeKind::Root => None,
            ScopeKind::Class(id, _) => Some(id.into()),
            ScopeKind::Def(id, _) => Some(id.into()),
            ScopeKind::Foreach(_, _) => None,
        }
    }

    pub fn record(&self) -> Option<&Record> {
        match &self.kind {
            ScopeKind::Root => None,
            ScopeKind::Class(_, record) => Some(record),
            ScopeKind::Def(_, record) => Some(record),
            ScopeKind::Foreach(_, _) => None,
        }
    }

    pub fn record_mut(&mut self) -> Option<&mut Record> {
        match &mut self.kind {
            ScopeKind::Root => None,
            ScopeKind::Class(_, record) => Some(record),
            ScopeKind::Def(_, record) => Some(record),
            ScopeKind::Foreach(_, _) => None,
        }
    }

    pub fn into_class(self) -> (ClassId, Record) {
        match self.kind {
            ScopeKind::Class(id, class) => (id, class),
            _ => panic!("not a class"),
        }
    }

    pub fn into_def(self) -> (DefId, Record) {
        match self.kind {
            ScopeKind::Def(id, def) => (id, def),
            _ => panic!("not a def"),
        }
    }

    pub fn add_variable(&mut self, name: EcoString, variable_id: VariableId) {
        self.name_to_variable.insert(name, variable_id);
    }

    pub fn find_variable(&self, name: &EcoString) -> Option<VariableId> {
        if let Some(var_id) = self.name_to_variable.get(name) {
            return Some(*var_id);
        }

        if let ScopeKind::Foreach(var_name, var_id) = &self.kind {
            if name == var_name {
                return Some(*var_id);
            }
        }

        None
    }
}
