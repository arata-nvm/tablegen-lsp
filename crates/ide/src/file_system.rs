use std::collections::{HashMap, VecDeque};
use std::path::{Path, PathBuf};

use ecow::EcoString;
use syntax::ast::AstNode;
use syntax::parser::{TextRange, TextSize};
use syntax::{SyntaxNode, SyntaxNodePtr, ast};

use crate::db::{Db, parse};

#[derive(Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq, Hash)]
pub struct FileId(pub u32);

#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
pub struct FilePosition {
    pub file: FileId,
    pub position: TextSize,
}

impl FilePosition {
    pub fn new(file: FileId, position: TextSize) -> Self {
        Self { file, position }
    }
}

#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
pub struct FileRange {
    pub file: FileId,
    pub range: TextRange,
}

impl FileRange {
    pub fn new(file: FileId, range: TextRange) -> Self {
        Self { file, range }
    }
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct FilePath(pub PathBuf);

impl FilePath {
    pub fn join(&self, path: impl AsRef<Path>) -> FilePath {
        FilePath(self.0.join(path))
    }

    pub fn parent(&self) -> Option<FilePath> {
        self.0.parent().map(FilePath::from)
    }
}

impl From<&Path> for FilePath {
    fn from(value: &Path) -> Self {
        Self(value.to_path_buf())
    }
}

#[derive(Debug, Default)]
pub struct FileSet {
    path_to_id: HashMap<FilePath, FileId>,
    id_to_path: HashMap<FileId, FilePath>,
}

impl FileSet {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn insert(&mut self, file_id: FileId, path: FilePath) {
        self.path_to_id.insert(path.clone(), file_id);
        self.id_to_path.insert(file_id, path.clone());
    }

    pub fn remove(&mut self, file_id: &FileId) {
        if let Some(path) = self.id_to_path.remove(file_id) {
            self.path_to_id.remove(&path);
        }
    }

    pub fn contains(&self, file_id: &FileId) -> bool {
        self.id_to_path.contains_key(file_id)
    }

    pub fn file_for_path(&self, path: &FilePath) -> Option<FileId> {
        self.path_to_id.get(path).copied()
    }

    pub fn path_for_file(&self, file_id: &FileId) -> &FilePath {
        &self.id_to_path[file_id]
    }

    pub fn iter_files(&self) -> impl Iterator<Item = FileId> + '_ {
        self.id_to_path.keys().copied()
    }
}

#[derive(Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq, Hash)]
pub struct SourceUnitId(FileId);

impl SourceUnitId {
    pub fn from_root_file(root_file: FileId) -> Self {
        Self(root_file)
    }
}

#[derive(Debug)]
pub struct SourceUnit {
    root: FileId,
    includes: HashMap<FileId, IncludeMap>,
}

impl SourceUnit {
    pub fn root(&self) -> FileId {
        self.root
    }

    pub fn iter_files(&self) -> impl Iterator<Item = FileId> + '_ {
        self.includes.keys().copied()
    }

    pub fn include_map(&self, file_id: &FileId) -> Option<&IncludeMap> {
        self.includes.get(file_id)
    }
}

pub trait FileSystem {
    fn assign_or_get_file_id(&mut self, path: FilePath) -> FileId;

    fn path_for_file(&self, file_id: &FileId) -> &FilePath;

    fn read_content(&self, file_path: &FilePath) -> Option<String>;
}

pub fn collect_sources<FS: FileSystem>(
    db: &mut dyn Db,
    fs: &mut FS,
    root: FileId,
    include_dirs: &[FilePath],
) -> SourceUnit {
    let mut includes = HashMap::new();

    let mut queue = VecDeque::new();
    queue.push_back(root);
    while let Some(file_id) = queue.pop_front() {
        if includes.contains_key(&file_id) {
            continue;
        }

        let parse = parse(db, file_id);

        let file_path = fs.path_for_file(&file_id);
        let file_dir = file_path.parent().expect("file dir not found");

        let mut include_dir_list = vec![file_dir];
        include_dir_list.extend(include_dirs.iter().cloned());

        let mut include_map = HashMap::new();
        for (include_id, include_path) in list_includes(parse.syntax_node()) {
            let Some(resolved_file_id) =
                resolve_include_file(db, fs, include_path, &include_dir_list)
            else {
                continue;
            };

            include_map.insert(include_id, resolved_file_id);
            queue.push_back(resolved_file_id);
        }
        includes.insert(file_id, include_map);
    }

    SourceUnit { root, includes }
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct IncludeId(pub SyntaxNodePtr);

pub type IncludeMap = HashMap<IncludeId, FileId>;

fn list_includes(root_node: SyntaxNode) -> Vec<(IncludeId, EcoString)> {
    (|| -> Option<_> {
        let source_file = ast::SourceFile::cast(root_node)?;
        let stmt_list = source_file.statement_list()?;
        let include_paths = stmt_list
            .statements()
            .filter_map(|stmt| match stmt {
                ast::Statement::Include(include) => {
                    let id = IncludeId(SyntaxNodePtr::new(include.syntax()));
                    let path = include.path()?.value();
                    Some((id, path))
                }
                _ => None,
            })
            .collect();
        Some(include_paths)
    })()
    .unwrap_or_default()
}

fn resolve_include_file<FS: FileSystem>(
    db: &mut dyn Db,
    fs: &mut FS,
    include_path: EcoString,
    include_dir_list: &[FilePath],
) -> Option<FileId> {
    for include_dir in include_dir_list {
        let candidate_file_path = include_dir.join(include_path.as_str());
        if let Some(file_content) = fs.read_content(&candidate_file_path) {
            let file_id = fs.assign_or_get_file_id(candidate_file_path.clone());
            db.set_file_content(file_id, file_content.as_str());
            return Some(file_id);
        }
    }
    None
}
