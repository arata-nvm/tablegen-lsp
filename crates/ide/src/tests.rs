use std::{collections::HashMap, path::Path, sync::Arc};

use crate::{
    db::{SourceDatabase, SourceDatabaseStorage},
    eval::EvalDatabaseStorage,
    file_system::{self, FileId, FilePath, FileSet, FileSystem},
};

const DEFAULT_FILE_PATH: &str = "/main.td";

pub fn single_file(fixture: &str) -> (TestDatabase, Fixture) {
    let mut f = Fixture::single_file(fixture);
    let db = TestDatabase::new(&mut f);
    (db, f)
}

pub fn multiple_files(fixture: &str) -> (TestDatabase, Fixture) {
    let mut f = Fixture::multiple_files(fixture);
    let db = TestDatabase::new(&mut f);
    (db, f)
}

#[salsa::database(SourceDatabaseStorage, EvalDatabaseStorage)]
#[derive(Default)]
pub struct TestDatabase {
    storage: salsa::Storage<Self>,
}

impl salsa::Database for TestDatabase {}

impl TestDatabase {
    fn new(f: &mut Fixture) -> Self {
        let mut db = Self::default();
        for (file_id, content) in f.files() {
            db.set_file_content(file_id, Arc::from(content));
        }

        let source_root = file_system::collect_sources(&mut db, f, f.root_file());
        db.set_source_root(Arc::new(source_root));

        db
    }
}

#[derive(Debug, Default)]
pub struct Fixture {
    file_contents: HashMap<FilePath, String>,
    file_ids: Vec<FileId>,

    file_set: FileSet,
    next_file_id: u32,
}

impl Fixture {
    fn single_file(fixture: &str) -> Self {
        let mut this = Self::default();
        this.insert_file(
            FilePath::from(Path::new(DEFAULT_FILE_PATH)),
            fixture.to_string(),
        );
        this
    }

    fn multiple_files(fixture: &str) -> Self {
        let mut this = Self::default();

        let mut lines = fixture
            .lines()
            .skip_while(|line| !line.starts_with("; "))
            .peekable();
        while let Some(line) = lines.next() {
            let path_str = line
                .strip_prefix("; ")
                .expect("expected header line to start with '; '");
            let path = FilePath::from(Path::new(path_str));

            let mut content = String::new();
            while let Some(line) = lines.peek() {
                if line.starts_with("; ") {
                    break;
                }

                let line = lines.next().expect("line must be present");
                content.push_str(line);
                content.push_str("\n");
            }

            this.insert_file(path, content);
        }

        this
    }

    fn files(&self) -> Vec<(FileId, &str)> {
        self.file_contents
            .iter()
            .map(|(path, content)| {
                (
                    self.file_set
                        .file_for_path(path)
                        .expect("file id must exist"),
                    content.as_str(),
                )
            })
            .collect()
    }

    fn root_file(&self) -> FileId {
        self.file_ids
            .first()
            .expect("at least one file must be exist")
            .clone()
    }

    fn insert_file(&mut self, path: FilePath, content: String) {
        let res = self.file_contents.insert(path.clone(), content);
        assert!(res.is_none(), "duplicate file path");

        let file_id = self.alloc_file_id();
        self.file_set.insert(file_id, path);
        self.file_ids.push(file_id);
    }

    fn alloc_file_id(&mut self) -> FileId {
        let file_id = FileId(self.next_file_id);
        self.next_file_id += 1;
        file_id
    }
}

impl FileSystem for Fixture {
    fn assign_or_get_file_id(&mut self, path: FilePath) -> FileId {
        match self.file_set.file_for_path(&path) {
            Some(file_id) => file_id,
            None => {
                let file_id = self.alloc_file_id();
                self.file_set.insert(file_id, path);
                file_id
            }
        }
    }

    fn path_for_file(&self, file_id: &FileId) -> Option<FilePath> {
        self.file_set.path_for_file(file_id)
    }

    fn read_content(&self, file_path: &FilePath) -> Option<String> {
        self.file_contents.get(file_path).cloned()
    }
}
