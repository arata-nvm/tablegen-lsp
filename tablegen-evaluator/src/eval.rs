use tablegen_parser::{
    ast::{self, AstNode},
    node::SyntaxNode,
};

use crate::{
    evaluator::Evaluator,
    record_keeper::{Record, RecordField, RecordKeeper, TemplateArg, Type, Value},
};

pub fn evaluate(root: SyntaxNode) -> Option<RecordKeeper> {
    let mut e = Evaluator::new();
    eval_file(root, &mut e);
    Some(e.finish())
}

fn eval_file(root: SyntaxNode, e: &mut Evaluator) {
    with(
        root.cast::<ast::File>()
            .and_then(|file| file.statement_list()),
        |statement_list| {
            for statement in statement_list.statements() {
                eval_statement(statement, e);
            }
        },
    );
}

fn eval_statement(statement: ast::Statement, e: &mut Evaluator) -> Option<()> {
    match statement {
        ast::Statement::Include(_) => todo!(),
        ast::Statement::Class(class) => eval_class(class, e),
        ast::Statement::Def(_) => todo!(),
        ast::Statement::Let(_) => todo!(),
    };
    None
}

fn eval_class(class: ast::Class, e: &mut Evaluator) {
    with(class.name().and_then(|id| id.value()), |name| {
        e.start_record(Record::new(name.clone()));

        with(class.template_arg_list(), |arg_list| {
            for arg in arg_list.args() {
                let Some(arg) = eval_template_arg(arg, e) else { continue; };
                e.add_record_template_arg(arg);
            }
        });

        with(class.record_body(), |record_body| {
            eval_record_body(record_body, e);
        });

        e.finish_record();
    });
}

fn eval_template_arg(arg: ast::TemplateArgDecl, e: &mut Evaluator) -> Option<TemplateArg> {
    let name = arg.name()?.value()?;
    let typ = eval_type(arg.r#type()?, e)?;
    let value = eval_value(arg.value()?, e)?;
    Some(TemplateArg::new(name.clone(), typ, value))
}

fn eval_record_body(record_body: ast::RecordBody, e: &mut Evaluator) {
    with(record_body.parent_class_list(), |class_list| {
        for parent_class in class_list.classes() {
            // TODO
        }
    });
    with(record_body.body(), |body| {
        for body_item in body.items() {
            eval_body_item(body_item, e);
        }
    });
}

fn eval_body_item(body_item: ast::BodyItem, e: &mut Evaluator) {
    match body_item {
        ast::BodyItem::FieldDef(def) => {
            if let Some(field) = eval_field_def(def, e) {
                e.add_record_field(field);
            }
        }
        ast::BodyItem::FieldLet(_) => todo!(),
        ast::BodyItem::Defvar(_) => todo!(),
        ast::BodyItem::Assert(_) => todo!(),
    }
}

fn eval_field_def(def: ast::FieldDef, e: &mut Evaluator) -> Option<RecordField> {
    let name = def.name()?.value()?;
    let typ = eval_type(def.r#type()?, e)?;
    let value = match def.value() {
        Some(value) => eval_value(value, e)?,
        None => Value::Uninitialized,
    };
    Some(RecordField::new(name.clone(), typ, value))
}

fn eval_type(typ: ast::Type, e: &mut Evaluator) -> Option<Type> {
    let typ = match typ {
        ast::Type::BitType(_) => Type::Bit,
        ast::Type::IntType(_) => Type::Int,
        ast::Type::StringType(_) => Type::String,
        ast::Type::DagType(_) => Type::Dag,
        ast::Type::BitsType(t) => Type::Bits(t.length()?.value()?),
        ast::Type::ListType(t) => Type::List(Box::new(eval_type(t.inner_type()?, e)?)),
        ast::Type::ClassId(t) => Type::Class(t.name()?.value()?.to_string()),
        ast::Type::CodeType(_) => Type::Code,
    };
    Some(typ)
}

fn eval_value(value: ast::Value, e: &mut Evaluator) -> Option<Value> {
    for inner_value in value.inner_values() {
        let value = eval_simple_value(inner_value.simple_value()?, e)?;
        for suffix in inner_value.suffixes() {
            match suffix {
                ast::ValueSuffix::RangeSuffix(_) => todo!(),
                ast::ValueSuffix::SliceSuffix(_) => todo!(),
                ast::ValueSuffix::FieldSuffix(_) => todo!(),
            }
        }
        return Some(value);
    }
    None
}

fn eval_simple_value(simple_value: ast::SimpleValue, e: &mut Evaluator) -> Option<Value> {
    let val = match simple_value {
        ast::SimpleValue::Integer(v) => Value::Integer(v.value()?),
        ast::SimpleValue::String(v) => Value::String(v.value()),
        ast::SimpleValue::Code(v) => Value::Code(v.value()?.to_string()),
        ast::SimpleValue::Boolean(v) => Value::Boolean(v.value()?),
        ast::SimpleValue::Uninitialized(_) => Value::Uninitialized,
        ast::SimpleValue::Bits(v) => Value::Bits(
            v.value_list()?
                .values()
                .map(|v| eval_value(v, e))
                .collect::<Option<Vec<Value>>>()?,
        ),
        ast::SimpleValue::List(v) => Value::List(
            v.value_list()?
                .values()
                .map(|v| eval_value(v, e))
                .collect::<Option<Vec<Value>>>()?,
        ),
        ast::SimpleValue::Dag(v) => todo!(),
        ast::SimpleValue::Identifier(v) => Value::Identifier(v.value()?.to_string()),
        ast::SimpleValue::ClassRef(v) => {
            let name = v.name()?.value()?.to_string();
            let positional_arg = v
                .arg_value_list()?
                .positional()?
                .values()
                .map(|v| eval_value(v, e))
                .collect::<Option<Vec<Value>>>()?;
            let named_arg = v
                .arg_value_list()?
                .named()?
                .values()
                .map(|v| Option::zip(eval_value(v.name()?, e), eval_value(v.value()?, e)))
                .collect::<Option<Vec<(Value, Value)>>>()?;
            Value::ClassRef(name, positional_arg, named_arg)
        }
        ast::SimpleValue::BangOperator(v) => {
            let op = v.kind()?;
            let args = v
                .values()
                .map(|v| eval_value(v, e))
                .collect::<Option<Vec<Value>>>()?;
            Value::BangOperator(op, args)
        }
        ast::SimpleValue::CondOperator(v) => {
            let clauses = v
                .clauses()
                .map(|c| Option::zip(eval_value(c.condition()?, e), eval_value(c.value()?, e)))
                .collect::<Option<Vec<(Value, Value)>>>()?;
            Value::CondOperator(clauses)
        }
    };
    Some(val)
}

fn with<A>(node: Option<A>, f: impl FnOnce(A)) {
    if let Some(node) = node {
        f(node);
    }
}
