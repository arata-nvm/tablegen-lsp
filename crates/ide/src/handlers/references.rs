use crate::eval::EvalDatabase;
use crate::file_system::{FilePosition, FileRange};

pub fn exec(db: &dyn EvalDatabase, pos: FilePosition) -> Option<Vec<FileRange>> {
    let evaluation = db.eval();
    let symbol_map = evaluation.symbol_map();
    let symbol = symbol_map.find_symbol_at(pos)?;
    Some(symbol.reference_locs().to_vec())
}

#[cfg(test)]
mod tests {
    use crate::tests;

    fn check(s: &str) -> Vec<String> {
        let (db, f) = tests::single_file(s);
        let references = super::exec(&db, f.marker(0)).expect("definition not found");
        let content = f.file_content(&f.root_file());
        references
            .into_iter()
            .map(|it| content[it.range].to_string())
            .collect()
    }

    #[test]
    fn class() {
        insta::assert_debug_snapshot!(check("class $Foo; class Bar: Foo;"));
        insta::assert_debug_snapshot!(check("class $Foo { Foo foo; }"));
    }

    #[test]
    fn template_arg() {
        insta::assert_debug_snapshot!(check("class Foo<int $arg1> { int field1 = arg1; }"));
    }

    #[test]
    fn field() {
        insta::assert_debug_snapshot!(check("class Foo { int $field1 = 1; int field2 = field1; }"));
    }
}
