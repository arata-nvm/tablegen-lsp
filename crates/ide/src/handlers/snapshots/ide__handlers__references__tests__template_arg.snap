---
source: crates/ide/src/handlers/references.rs
expression: "check(\"class Foo<int $arg1> { int field1 = arg1; }\")"
---
[
    "arg1",
]
